luaa.c

```diff
@@ -411,6 +412,126 @@ get_modifier_name(int map_index)
+    }
+    else
+        return 0;
+    for (size_t i = 1; i < length; i++)
+    {
+        if ((*(input + i) & 0x000000c0) != 0x00000080)
+            return 0;
+        ucs += (*(input + i) & 0x0000003f) << (6 * (length-i-1));
+    }
+    return ucs;
+}
+
+/** Get Xlib keysym and xkb_keysym_to_uf8 from an Awesome keycode.
+ *
+ * xkb_keysym_to_utf8 returns a UTF-8 representation of the symbol a given
+ * symkey generates. It matches the engraving of the key for level-0 symbols
+ * (but lowercase)
+ *
+ * Keycodes may be given in a string in any valid format for `awful.key`:
+ * "#" + keycode, the symkey name and the UTF-8 representation will all work.
+ *
+ * If no suitable keysym is found, or a malformed keycode is given as an
+ * argument, this function will return (nil, nil)
+ *
+ * @treturn[1] string keysym The keysym name
+ * @treturn[1] nil keysym If no valid keysym is found
+ * @treturn[2] string printsymbol The xkb_keysym_to_utf8 result
+ * @treturn[2] nil printsymbol If the keysym has no printable representation.
+ * @staticfct get_key_name
+ */
+
+static int
+luaA_get_key_name(lua_State *L)
+{
+    // check if argument is valid
+    if (lua_gettop(L) > 1 || lua_type(L, 1) != LUA_TSTRING)
+    {
+        return 0;
+    }
+
+    const char* input = luaL_checkstring(L, 1);
+    const xkb_keysym_t *keysyms;
+    xkb_keysym_t keysym = XKB_KEY_NoSymbol;
+    const size_t length = strlen(input);
+    uint32_t ucs;
+
+    /* Checking for the three possible syntaxes awful.key uses:
+     * 1: #keycode (#8 to #255, any other is invalid)
+     * 2: the symbol itself (the result of xkb_keysym_to_utf8, e.g. @ for at).
+     * 3: the keysym
+     */
+    if (length > 1 && *input == '#' && length < 5) // syntax #1
+    {
+        int keycode_from_hash = 0;
+        for (size_t i = 1; i < length; i++)
+        {
+            if ('9' < *(input + i) || *(input + i) < '0')
+                return 0;
+            else
+            {
+                keycode_from_hash *= 10;
+                keycode_from_hash += input[i] - 48;
+            }
+        }
+        // A keycode cannot go lower than 8 or higher than 255
+        if ((keycode_from_hash < 8) || keycode_from_hash > 255)
+            return 0;
+        xkb_keycode_t keycode = (xkb_keycode_t) keycode_from_hash;
+        struct xkb_keymap *keymap = xkb_state_get_keymap(globalconf.xkb_state);
+        xkb_keymap_key_get_syms_by_level(keymap, keycode, 0, 0, &keysyms);
+        keysym = keysyms[0];
+    }
+    else if ((ucs = one_utf8_to_utf32(input, length)) > 0) //syntax #2
+        keysym = xkb_utf32_to_keysym(ucs);
+    else //syntax #3
+        keysym = xkb_keysym_from_name(input, XKB_KEYSYM_NO_FLAGS);
+
+    if (keysym == XKB_KEY_NoSymbol)
+    {
+        lua_pushnil(L);
+        lua_pushnil(L);
+    }
+    else
+    {
+        char *name = key_get_keysym_name(keysym);
+        lua_pushstring(L, name);
+        char *utfname;
+        utfname=malloc(8);
```

Comment:
That's a clear memory leak. Just do (how do I make a suggestion to replace two lines with one on GitHub?):
```suggestion
        char utfname[8];
```

Reply:
Thank you. I'll ask you if I can't make it work. (The compiler was giving me some warnings about passing a pointer instead of a value or something like that...)

Reply:
In the "worst case", you have to pass in `&utfname[0]` instead of `utfname` to the function in the next line.

Reply:
Okay, done. No problems with `char utfname[8]`.

By the way, I rolled all the fixes you suggested in the same commit. I can rebase later if you all want a cleaner history.